ts/ui/lazy/LazyHandler.ts

```diff
@@ -0,0 +1,483 @@
+     *
+     * @param {IntersectionObserverEntry[]} entries   The markers that have come into or out of view.
+     */
+    protected lazyObserve(entries: IntersectionObserverEntry[]) {
+      for (const entry of entries) {
+        const id = this.adaptor.getAttribute(entry.target as any as N, LAZYID);
+        const math = this.lazyList.get(id);
+        if (!math) return;
+        if (entry.isIntersecting) {
+          this.lazySet.add(id);
+          if (!this.lazyIdle) {
+            this.lazyIdle = true;
+            this.lazyProcessSet();
+          }
+        } else {
+          this.lazySet.delete(id);
+        }
+      }
+    }
+
+    /**
+     * Mark the MathItems in the set as needing compiling or typesetting,
+     *   and for TeX items, make sure the earlier TeX items are typeset
+     *   (in case they have automatic numbers, or define macros, etc.).
+     * Then rerender the page to update the visible equations.
+     */
+    protected lazyHandleSet() {
+      const set = this.lazySet;
+      this.lazySet = new Set();
+      this.lazyPromise = this.lazyPromise.then(() => {
+        let state = this.compileEarlierItems(set) ? STATE.COMPILED : STATE.TYPESET;
+        state = this.resetStates(set, state);
+        this.state(state - 1, null); // reset processed bits to allow reprocessing
+        return handleRetriesFor(() => {
+          this.render();
+          this.lazyIdle = false;
+        });
+      });
+    }
+
+    /**
+     * Set the states of the MathItems in the set, depending on
+     *   whether they need compiling or just typesetting, and
+     *   update the state needed for the page rerendering.
+     *
+     * @param {LazySet} set    The set of math items to update
+     * @param {number} state   The state needed for the items
+     * @return {number}        The updated state based on the items
+     */
+    protected resetStates(set: LazySet, state: number): number {
+      for (const id of set.values()) {
+        const math = this.lazyList.get(id);
+        if (math.lazyCompile) {
+          math.state(STATE.COMPILED - 1);
+          state = STATE.COMPILED;
+        } else {
+          math.state(STATE.TYPESET - 1);
+        }
+        math.lazyCompile = math.lazyTypeset = false;
+        math.lazyMarker && this.lazyObserver.unobserve(math.lazyMarker as any as Element);
+      }
+      return state;
+    }
+
+    /**
+     * Mark any TeX items (earlier than the ones in the set) to be compiled.
+     *
+     * @param {LazySet} set   The set of items that are newly visible
+     * @return {boolean}      True if there are TeX items to be typeset
+     */
+    protected compileEarlierItems(set: LazySet): boolean {
+      let math = this.earliestTex(set);
+      if (!math) return false;
+      let compile = false;
+      for (const item of this.math) {
+        if (item === math) break;
+        const earlier = item as LazyMathItem<N, T, D>;
+        if (earlier.lazyTex && earlier.lazyCompile) {
```

Comment:
Could you cast the list to `LazyMathItem<>[]`, run the loop on `item` only and  have a single guard we you break?
Would typescript be unhappy with that?

Reply:
No, because `this.math` is not an array, though it is iterable.  One could use `Array.from()` to make it an array, but since there could be hundreds (ben thousands) of equations you will be creating large arrays every time an expression is rendered.  The whole point of the iterable linked list for `this.math` was to be able to iterate it without needing to make it into an array.

It would be nice to be able to cast `this.math` to a list of `LazyMathItem<N, T, D>`, but the types are set up right for that because `this.math` is a `MathList<N, T, D>`, and that is based on `MathItem<N, T, D>`.  If you want, I could refactor that so that `MathList` has one generic (the item type), and we use `MathList<MathItem<N, T, D>>` instead of `MathList<N, T, D>`, so that this look could be

```
for (const item of this.math as `MathList<LazyMathItem<N, T, D>>) {...}
```

I haven't tried it, but I think that would work.

Alternatively, I could at least do 

```
        const earlier = item as LazyMathItem<N, T, D>;
        if (earlier === math || !earlier?.lazyCompile) {
         break;
        }
```

Which would you prefer?

Reply:
Version 1 is seems overkill to avoid a single recurring cast. So the alternative is fine.

Reply:
OK, made the change.